File: examples/lh-server/typescript/src/11-correlated-events/main.ts
import { createTaskWorker, LHConfig, Workflow } from "littlehorse-client";
import { z } from "zod";
import { closeOnShutdown } from "../config.js";

function processSignedDocument(documentId: string, signerName: string): string {
  const result = `Document ${documentId} was signed by ${signerName}. Processing complete.`;
  console.log(result);
  return result;
}

const config = LHConfig.from({});
const worker = createTaskWorker(
  processSignedDocument,
  "processSignedDocument",
  config,
  {
    inputVars: { documentId: z.string(), signerName: z.string() },
    outputSchema: z.string(),
  },
);

const workflow = Workflow.newWorkflow("correlated-event-example", (wf) => {
  const documentId = wf.declareStr("document-id");
  const signerName = wf.waitForEvent("document-signed", {
    correlationId: documentId,
    correlatedEventConfig: { deleteAfterFirstCorrelation: true },
    payloadSchema: z.string(),
  });
  wf.execute("processSignedDocument", documentId, signerName);
});

await closeOnShutdown([worker]);
await worker.registerTaskDef();
await workflow.registerWfSpec(config);
console.log("Run with: lhctl run correlated-event-example document-id my-document-abc123");
console.log("Then post with: npm run correlated-events:post");
await worker.start();

File: examples/lh-server/go/11-correlated-events/main.go
package main

import (
	"github.com/littlehorse-enterprises/lh-developer-hub/examples/lh-server/go/internal/examplesupport"
	"github.com/littlehorse-enterprises/littlehorse/sdk-go/littlehorse"
)

func main() {
	config := littlehorse.NewConfigFromEnv()
	worker := examplesupport.NewWorker(config, ProcessSignedDocument, "processSignedDocument")
	workflow := littlehorse.NewWorkflow(WfLogic, "correlated-event-example")
	examplesupport.Start(config, []*littlehorse.LHTaskWorker{worker}, workflow)
}
